File: python/Backtracking/78.subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res: List[List[int]] = []
        n = len(nums)

        def backtrack(first: int, curr: List[int]):
            if len(curr) == k:
                res.append(curr)
                return
            for i in range(first, n):
                backtrack(i + 1, curr + [nums[i]])

        for k in range(n + 1):
            backtrack(0, [])
        return res

File: TypeScript/Backtracking/78.subsets.ts
function subsets(nums: number[]): number[][] {
  let res: number[][] = []
  let subset: number[] = []
  let n = nums.length
  dfs(0)
  return res
  function dfs(i: number) {
    if (i >= n) {
      res.push([...subset])
      return
    }
    subset.push(nums[i])
    dfs(i + 1)
    subset.pop()
    dfs(i + 1)
  }
}
function subsets1(nums: number[]): number[][] {
  let res = []
  backtrack(0, [])

  function backtrack(start: number, cur: number[]) {
    res.push([...cur])
    for (let i = start; i < nums.length; ++i) {
      backtrack(i + 1, [...cur, nums[i]])
    }
  }

  return res
}
console.log(subsets1([1, 2, 3]))
